pkg/core/sync.go

```diff
@@ -0,0 +1,210 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/cloudquery/cloudquery/internal/logging"
+	"github.com/cloudquery/cloudquery/pkg/core/database"
+	"github.com/cloudquery/cloudquery/pkg/core/state"
+	"github.com/cloudquery/cloudquery/pkg/plugin"
+	"github.com/cloudquery/cloudquery/pkg/plugin/registry"
+	sdkdb "github.com/cloudquery/cq-provider-sdk/database"
+	"github.com/cloudquery/cq-provider-sdk/migration"
+	"github.com/cloudquery/cq-provider-sdk/migration/migrator"
+	"github.com/cloudquery/cq-provider-sdk/provider/diag"
+	"github.com/cloudquery/cq-provider-sdk/provider/execution"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+	"github.com/rs/zerolog/log"
+	"github.com/thoas/go-funk"
+)
+
+type SyncState int
+
+const (
+	Upgraded SyncState = iota + 1
+	Downgraded
+	NoChange
+)
+
+func (s SyncState) String() string {
+	switch s {
+	case Upgraded:
+		return "Upgraded"
+	case Downgraded:
+		return "Downgraded"
+	case NoChange:
+		return "NoChange"
+	}
+	return "Unknown"
+}
+
+type SyncOptions struct {
+	Provider registry.Provider
+	// DownloadLatest specifies whether the latest provider should be downloaded before sync is called.
+	DownloadLatest bool
+}
+
+type SyncResult struct {
+	State      SyncState
+	OldVersion string
+	NewVersion string
+}
+
+func Sync(ctx context.Context, storage database.Storage, pm *plugin.Manager, opts *SyncOptions) (*SyncResult, diag.Diagnostics) {
+	log.Info().Stringer("provider", opts.Provider).Msg("syncing provider schema")
+	if opts.DownloadLatest {
+		if _, diags := Download(ctx, pm, &DownloadOptions{
+			[]registry.Provider{{Name: opts.Provider.Name, Version: registry.LatestVersion, Source: opts.Provider.Source}}, false}); diags.HasErrors() {
+			return nil, diags
+		}
+	}
+
+	rp := registry.Provider{Name: opts.Provider.Name, Version: registry.LatestVersion, Source: opts.Provider.Source}
+
+	// always use latest available provider available
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: rp})
+	if diags.HasDiags() {
+		return nil, diags
+	}
+
+	// TODO: in future more components will want state, so make state more generic and passable via database.Storage
+	db, err := sdkdb.New(ctx, logging.NewZHcLog(&log.Logger, "sync"), storage.DSN())
+	if err != nil {
+		return nil, diag.FromError(err, diag.DATABASE)
+	}
+	defer db.Close()
+
+	rp.Version = s.Version
+
+	want := state.ProviderFromRegistry(rp)
+	if want.ParsedVersion == nil {
+		return nil, diag.FromError(fmt.Errorf("failing provider with invalid version %q", rp.Version), diag.INTERNAL)
+	}
+
+	sta, err := state.NewMigratedClient(ctx, storage.DSN(), logging.NewZHcLog(&log.Logger, "sync"))
```

Comment:
I don't see you call in all places NewMigratedClient, why not just call it NewClient?

Reply:
it wouldn't migrate core.

Reply:
I think we should always migrate core when creating the client, so it doesn't really matter to give it a special name.

Reply:
but then policy stuff doesn't work because of interfaces.

Reply:
it will be fixed, you shouldn't think too hard about this issue